src/main/scala/se/lu/nateko/cp/meta/services/sparql/magic/index/IndexData.scala

```diff
@@ -272,14 +335,122 @@ final class IndexData(nObjects: Int)(
+				updateProjectKeywords(subj, isAssertion, projectKeywords)
+
+			case `hasKeywords` =>
+				getDataObject(subj) match {
+					case Some(oe) => {
+						obj.asOptInstanceOf[Literal].flatMap(asString).flatMap(s => Some(parseCommaSepList(s))).toSeq.flatten.foreach { strVal =>
+							updateCategSet(categMap(Keyword), strVal, oe.idx, isAssertion)
+						}
+					}
+					case None => {
+						val changedKeywords = parseCommaSepList(obj.stringValue()).toSet
+						val isSpec = StatementSource.hasStatement(null, vocab.hasObjectSpec, subj)
+
+						// TODO: If we can assume specs are always inserted before their hasKeywords triple,
+						//			 we could use the `specs` array to know if this is a spec.
+						if (isSpec) {
+							updateSpecKeywords(ensureIRI(subj), isAssertion, changedKeywords)
+						} else {
+							val project = subj
+							StatementSource.getStatements(null, vocab.hasAssociatedProject, project)
+								.map(_.getSubject())
+								.foreach(spec =>
+									// TODO: Does not cover the case where two projects have overlapping keywords
+									updateProjectKeywords(ensureIRI(spec), isAssertion, changedKeywords)
+								)
+						}
+					}
 				}
 
 			case _ =>
 		}
 	}
 
+	private def updateSpecKeywords(spec: IRI, isAssertion: Boolean, changedSpecKeywords: Set[String])(using
+		vocab: CpmetaVocab
+	)(using StatementSource): Unit = {
+		val existingKeywords: Set[String] = getKeywords(spec)
+		val projKeywords = getSpecProjectKeywords(spec)
+		val newKeywords: Set[String] = projKeywords ++ modifyKeywords(isAssertion, existingKeywords, changedSpecKeywords)
+
+		setSpecKeywords(spec, newKeywords)
+	}
+
+	private def updateProjectKeywords(spec: IRI, isAssertion: Boolean, changedProjectKeywords: Set[String])(using
+		vocab: CpmetaVocab
+	)(using StatementSource): Unit = {
+		val projKeywords = getSpecProjectKeywords(spec)
+		val newKeywords: Set[String] = getKeywords(spec) ++ modifyKeywords(isAssertion, projKeywords, changedProjectKeywords)
+
+		setSpecKeywords(spec, newKeywords)
+	}
+
+	private def modifyKeywords(isAssertion: Boolean, existing: Set[String], changed: Set[String]): Set[String] = {
+		if (isAssertion) {
+			existing ++ changed
+		} else {
+			existing -- changed
+		}
+	}
+
+	private def getSpecProjectKeywords(spec: IRI)(using vocab: CpmetaVocab)(using StatementSource): Set[String] = {
+		if (spec == null) {
+			return Set.empty;
+		}
+
+		StatementSource.getUriValues(spec, vocab.hasAssociatedProject)
+			.flatMap(project => getKeywords(project))
+			.toSet
+	}
+
+	private def ensureIRI(value: Value): IRI = {
+		value match {
+			case iri: IRI => iri
+		}
+	}
+
+	private def getKeywords(subject: IRI)(using vocab: CpmetaVocab)(using StatementSource): Set[String] = {
+		if (subject == null) {
+			return Set.empty;
+		}
+
+		StatementSource.getValues(subject, vocab.hasKeywords)
+			.flatMap(parseKeywords)
+			.flatten()
+			.toSet
+	}
+
+	private def setSpecKeywords(spec: IRI, keywords: Set[String]) = {
```

Comment:
This method also does not have to deal with bitmaps if we simplify the meaning of `_keywordsToSpecs` as suggested above.